### ISSUE TITLE
Ignore isn't using .gitignore or .ignore

### ISSUE THREAD

**Issue by matanmarkind** [watchexec/watchexec#250]
Hi,

I am currently running watchexec 1.18.4 on ubuntu 20.04.

I use pytest, which makes changes to /.pytest_cache/. I am ignoring this in .gitignore (`**pytest_cache**`), but this doesn't seem to get picked up by watchexec. I eventually found that using `watchexec -i ".pytest_cache/**"` does work.

Adding the working glob to .gitignore or to .ignore are not picked up by watchexec though. I did not pass --no-project-ignore. It is also surprising that the glob which works in .gitignore doesn't work ehre.

The command I ran which does ignore is:
`watchexec -i ".pytest_cache/**" --print-events -- pytest exercise-4/q2/tests/test_listener.py`

This prints events to the screen showing what changes cause the loop (AFAICT):
[EVENT 0] Event source=Filesystem kind=Modify(Data(Any)) path=/home/matan/Documents/gittick-course/.pytest_cache/v/cache/stepwise filetype=file
[EVENT 1] Event source=Filesystem kind=Access(Close(Write)) path=/home/matan/Documents/gittick-course/.pytest_cache/v/cache/stepwise filetype=file
[EVENT 2] Event source=Filesystem kind=Modify(Data(Any)) path=/home/matan/Documents/gittick-course/.pytest_cache/v/cache/nodeids filetype=file
[EVENT 3] Event source=Filesystem kind=Modify(Data(Any)) path=/home/matan/Documents/gittick-course/.pytest_cache/v/cache/nodeids filetype=file
[EVENT 4] Event source=Filesystem kind=Access(Close(Write)) path=/home/matan/Documents/gittick-course/.pytest_cache/v/cache/nodeids filetype=file
[EVENT 5] Event source=Internal command-completed(ExitError(1))

The output of -vvv is too big for me to paste here (>3k lines with ignore) so I will send that to you as a file.

**Comment by passcod** [comment]
Right, I've looked at the logs, and there's something interesting happening which I think is an issue with globset. Relevant section, edited to be readable:

```yaml
watchexec::action: got event
  event=Event { tags: [
    Source(Filesystem)
    FileEventKind(Modify(Data(Any)))
    Path { path: "/home/matan/Documents/gittick-course/.pytest_cache/v/cache/stepwise", file_type: Some(File)
  } ], metadata: {} }

filterer_check: watchexec::filter::globset: checking internal ignore filterer
filterer_check:filterer_check:checking_against_compiled:
  watchexec::ignore::filter: checking against path or parents
  watchexec::ignore::filter: positive match, but not in scope (ignore)
    glob=Glob {
      from: Some("/home/matan/Documents/gittick-course/exercise-2/.env")
      original: "*"
      actual: "**/*"
      is_whitelist: false
      is_only_dir: false
    }
  watchexec::ignore::filter: verdict pass=true
```

When `-i` is passed, this section is followed by:

```yaml
  watchexec::filter::globset: ignored by globset ignore

watchexec::action: filter rejected event
```

Walking through:

- `watchexec::ignore::filter` deals exclusively in ignore files, `watchexec::filter::globset` is the bit that treats with the `-i`, `-f`, `-e` options. So the first section (and the 2000 lines above it which deal with ignore file discovery) shows us the ignore files _are_ being loaded, and are being used.
- For dealing with ignore files, we delegate almost entirely to the [ignore](https://crates.io/crates/ignore) crate, which itself uses the [globset](https://crates.io/crates/globset) crate. These are both part of [ripgrep](https://github.com/BurntSushi/ripgrep).
- Here, there's a gitignore at `exercise-2/.env/.gitignore` which contains the glob `*`. That glob is scoped to `exercise-2/.env`, but `ignore` is picking this up and returning it as _the_ match to the event path. A positive match here would mean "ignore the event."
- However, in post (watchexec code, not ignore code) we check the `from` bit against the path and (rightfully!) reject that verdict. That glob should not have matched, as it is out of scope. So the verdict changes to pass.
- With `-i`, the ignore/filter/extension patterns are matched after this process, and correctly fail the event.

The issue:

1. The ignore crate cannot give us all the matches. So we're stuck considering only the first one it gives us.
2. The ignore crate doesn't limit its globs to those that are in scope for the tested path.
3. Thus, in this case where a _very broad_ ignore is loaded that should be out of scope, the ignore files are rendered essentially useless.

Further thoughts:

- It's unclear whether this is a bug in the ignore crate. As it's a component of ripgrep, its design is rather restricted to whatever makes sense in a grep context. It may be that the current behaviour is perfectly fine in the ripgrep usecase, so there wouldn't be any luck getting it "fixed." More investigation needed.
- There's no easy fix to this. If the ignore crate is "working as designed," I'll need to write a custom equivalent of the ignore crate that is instead geared towards the watchexec context. If it's a bug in the ignore crate, it will need fixing there, which has an unknown complexity to it (i.e I'm not at all familiar with it).
- It may have "worked" in 1.17 and below but mostly by chance (shadowed by other bugs) if it even did.

Workaround:

(This is very ugly, impractical, and it's not really a good long term solution. But in the immediate, it should work!)

- Pass `--no-project-ignore`
- Set the `WATCHEXEC_IGNORE_FILES` environment variable to the path of the main `.gitignore` you want (more ignores can be separated by `:`).
- That way, only that one project ignore file will be loaded.

In your case, it may look a bit like this:

```console
$ env WATCHEXEC_IGNORE_FILES=.gitignore watchexec --no-project-ignore -- pytest exercise-4/q2/tests/test_listener.py
```

**Comment by matanmarkind** [comment]
Cool thanks for the explanation and solution. I actually just added to my .bashrc file:
`alias wexec="WATCHEXEC_IGNORE_FILES=.gitignore watchexec --no-project-ignore"`
So now I can run
```
wexec -- pytest exercise-4/q2/tests/test_listener.py
```
Which seems to work fine.  Obviously this has the downside of putting a layer between the user and watchexec, but it's easier to type/read :)

**Comment by passcod** [comment]
1.22.3 should have fixed this.
